src/circuit/v2/hop.ts

```
@@ -0,0 +1,219 @@
+  } catch (e: any) {
+    log.error('error passing reserve message response from %s because', connection.remotePeer, e.message)
+    stream.close()
+    throw e
+  }
+
+  if (response.status === Status.OK && (response.reservation != null)) {
+    return response.reservation
+  }
+  const errMsg = `reservation failed with status ${response.status ?? 'undefined'}`
+  log.error(errMsg)
+  throw new Error(errMsg)
+}
+
+const isRelayAddr = (ma: Multiaddr): boolean => ma.protoCodes().includes(CIRCUIT_PROTO_CODE)
+
+async function handleReserve ({ connection, stream: pbstr, relayPeer, relayAddrs, limit, acl, reservationStore, peerStore }: HopProtocolOptions): Promise<void> {
+  const hopstr = pbstr.pb(HopMessage)
+  log('hop reserve request from %s', connection.remotePeer)
+
+  if (isRelayAddr(connection.remoteAddr)) {
+    log.error('relay reservation over circuit connection denied for peer: %p', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  if ((await acl?.allowReserve?.(connection.remotePeer, connection.remoteAddr)) === false) {
+    log.error('acl denied reservation to %s', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  const result = await reservationStore.reserve(connection.remotePeer, connection.remoteAddr)
+
+  if (result.status !== Status.OK) {
+    hopstr.write({ type: HopMessage.Type.STATUS, status: result.status })
+    return
+  }
+
+  try {
+    // tag relayed peer
+    // result.expire is non-null if `ReservationStore.reserve` returns with status == OK
+    if (result.expire != null) {
+      const ttl = new Date().getTime() - result.expire
+      await peerStore.tagPeer(relayPeer, RELAYED, { value: 1, ttl })
+        .catch((err: CodeError) => {
+          // ignore if peer is already tagged
+          if (err.code !== 'ERR_DUPLICATE_TAG') {
+            throw err
+          }
+        })
```

If the peer is already tagged, this should update the TTL of the tag if it's going to expire before the reservation rather than swallowing the error.

Also, I can't see any tests for this?